IPNI_FEDERATION.md

```diff
@@ -0,0 +1,406 @@
+# IPNI Federation Protocol
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+**Author(s)**:
+
+- [Masih Derkani](https://github.com/masih)
+
+**Maintainer(s)**:
+
+- [Andrew Gillis](https://github.com/gammazero)
+- [Ivan Schasny](https://github.com/ischasny)
+- [Masih Derkani](https://github.com/masih)
+- [Will Scott](https://github.com/willscott)
+
+* * *
+
+**Abstract**
+
+This document provides an overview of the IPNI federation protocol: a protocol by which a set of IPNI instances
+collaborate in order to provide a unified eventually consistent view of the index records.
+
+## Introduction
+
+The Inter-Planetary Network Indexer (IPNI) offers a routing system that enables mass advertisement of content addressed
+data and lookup performance in order of milliseconds.
+This is achieved by a design where a single IPNI instance strives to maintain full network state knowledge.
+Many IPNI instances can be instantiated across the globe to offer fast local ingestion and lookup of advertised content.
```

Comment:
Another goal of federation may be to allocate portions of the index content to different indexers. I think this should be talked about - even if only to say why we do not do it. People will want to index some stuff for some time, but not everything forever, and will want to be part of a federation for some of the other features.

Having local fast lookup is a good goal, but does that apply to everything? It may not be necessary to be fast and local for all storage providers, but just the highly accessed ones. This means we may want to consider a strategy that allows some subset of service providers, and even epochs within those providers' chains, to be assigned to various indexers. Infrequently accessed storage provider content can be indexed by fewer indexers.

Such a strategy may also be useful so that indexers can choose which providers and epochs are the best to index and not have to index everything, even from specific provider(s).

Reply:
Though useful, this feature seems sufficiently complicated to need its own specification?

Aside from that, I think what's documented in the current spec doesn't technically limit that behaviour. For example, there is nothing that would stop a group of collaborating indexers to agree on only indexing advertisements up to the depth 1000 within the last week. The federation protocol described here would still work as long as all the participating nodes only include providers with such criteria. 

WDYT @gammazero ?

Reply:
I suppose that is the purpose of this spec - to make indexing consistent between indexers that cover identical index content.  It means that if indexer-1 covers providers A and B, indexer-2 covers B and C, and indexer-3 covers C and A, these cannot be in the same federation even though they can check each other's consistency. Or, are they each part of 2 federations?

I can see index allocation, and incentivization as completely different specs.

Reply:
> I suppose that is the purpose of this spec

It all boils down to the words "Collaborating IPNI indexers".
The current spec aggregates provider lists across all indexers.  This means one indeed cannot have an indexer that only indexes a proportion of providers instead of all them. 

In the case you described, at least 2 of those indexers need to agree on the heuristic for selecting providers. Right now that heuristic in cid.contact is simply "All that you can get your hands on but not the ones you have not seen for a week".  If indexer-1, 2 and 3 follow the current heuristic and are part of the same federation, then the should all end up with A, B and C in their providers list.

The specification of what that heuristic is, IMO is beyond the scope of this spec. I think we can both imagine case where it can grow quite complicated, and could potentially lead into a lot of head scratching for users trying to look things up.

Considering the spec does not prohibit the federation of a set of indexers that agree upon the same heuristic to select providers, my vote would be to keep the scope small and aim for a federation of cid.contact-like indexers going first.

In that world, one could use DNS names to form separate federations across indexers that only care about "hot off the wire" records. For example, today.cid.contact could be a new indexer part of a separate federation of indexers that only index ads published within the last 24 hours.